internal/patroni/config.go

```diff
@@ -227,17 +227,24 @@ func DynamicConfiguration(
 	// Override the above with mandatory parameters.
 	if pgParameters.Mandatory != nil {
 		for k, v := range pgParameters.Mandatory.AsMap() {
-			// Unlike other PostgreSQL parameters that have mandatory values,
-			// shared_preload_libraries is a comma separated list that can have
-			// other values appended in addition to the mandatory values. Below,
-			// any values provided in the CRD are appended after the mandatory
-			// values.
-			s, ok := parameters[k].(string)
-			if k == "shared_preload_libraries" && ok {
-				parameters[k] = v + "," + s
-			} else {
-				parameters[k] = v
+
+			// This parameter is a comma-separated list. Rather than overwrite the
+			// user-defined value, we want to combine it with the mandatory one.
+			// Some libraries belong at specific positions in the list, so figure
+			// that out as well.
+			if k == "shared_preload_libraries" {
+				// Load mandatory libraries ahead of user-defined libraries.
+				if s, ok := parameters[k].(string); ok && len(s) > 0 {
+					v = v + "," + s
+				}
+				// Load "citus" ahead of any other libraries.
+				// - https://github.com/citusdata/citus/blob/v12.0.0/src/backend/distributed/shared_library_init.c#L417-L419
+				if strings.Contains(v, "citus") {
+					v = "citus," + v
```

Comment:
Making sure I've got this (though the test helps): 

1. we want citus to be first; 
2. citus can be present several times without side effect;
3. removing citus from that list (from the non-first position) would be brittle

Is that a fair summation?

Reply:
> Is that a fair summation?

Yep.

> 2\. citus can be present several times without side effect;

There is a set of related parameters for loading libraries that amount to `LOAD` statements at certain times.

- https://www.postgresql.org/docs/current/runtime-config-client.html#RUNTIME-CONFIG-CLIENT-PRELOAD
- https://www.postgresql.org/docs/current/sql-load.html
  > If the file has been loaded already, the command does nothing.

Postgres [compares inodes](https://git.postgresql.org/gitweb/?p=postgresql.git;f=src/backend/utils/fmgr/dfmgr.c;hb=REL_11_0#l169) to detect symlink chains as well.

> 3\. removing citus from that list (from the non-first position) would be brittle

A variety of names can refer to the same library. As I recall, `citus`, `citus.so`, and a full path are acceptable. This does not handle the full path case, but in my experience that is quite unusual.

Being more deliberate with these names requires parsing, which #3530 did partially. Ultimately, matching on names is still a step removed from the issue: the Citus code that complains could be in a file with any name. :neckbeard: